ir/type.h

```diff
@@ -123,9 +128,38 @@ class VectorType final : public Type {
 };
 
 
+// Currently only supports aggregate type with two children
+// of int type.
+// Eg: {ix, i1}. Enforced in getTypeConstraints()
+class AggregateType final : public Type {
+  std::vector<Type*> childrenType;
+
+public:
+  AggregateType(std::string &&name) : Type(std::move(name)) {}
+  AggregateType(std::string &&name, std::initializer_list<Type*> types)
+    : Type(std::move(name)), childrenType(types) {
+    ENSURE(!childrenType.empty());
```

Comment:
it can be empty, no problem.
Also, ENSURE != assert.

Reply:
hmm, my understanding is that assert() only asserts if NDEBUG is not defined. Looking at definition of ENSURE in compiler.h, it also says the same: if NDEBUG is not defined, it asserts, otherwise it's a nop. So then, isn't writing a ENSURE or assert same thing then?

I also think that definition of ENSURE in compiler.h right now is superfluous as assert is going to be nop anyways if NDEBUG is defined - we don't need to specifically take care of that.

Reply:
ENSURE(x) always executes 'x'. It guarantees side-effects, while assert doesn't. That's the difference.